File: src/commands/builds.ts
import type { Command } from "commander";
import { resolveContext } from "../cli.ts";
import { printOutput } from "../output.ts";
import { followConsole } from "../console.ts";

export function registerBuildCommands(program: Command): void {
  program
    .command("console")
    .description("Get console output for a build")
    .argument("<name>", "job name")
    .argument("[build]", "build number (default: lastBuild)", "lastBuild")
    .option("--follow", "follow output in real-time", false)
    .action(async (name: string, build: string, opts: { follow: boolean }) => {
      const ctx = await resolveContext(program);
      const jobUrl = ctx.client.jobUrl(name);

      if (opts.follow) {
        await followConsole(ctx.client, jobUrl, build);
      } else {
        const text = await ctx.client.getText(`${jobUrl}/${build}/consoleText`);
        process.stdout.write(text);
      }
    });

  program
    .command("get-build")
    .description("Get build details (status, result, duration, etc.)")
    .argument("<name>", "job name")
    .argument("[build]", "build number (default: lastBuild)", "lastBuild")
    .action(async (name: string, build: string) => {
      const ctx = await resolveContext(program);
      const data = await ctx.client.getJson<BuildInfo>(
        `${ctx.client.jobUrl(name)}/${build}/api/json?tree=number,result,building,duration,estimatedDuration,timestamp,displayName,description,url`,
      );
      if (ctx.json) {
        printOutput(data, true);
      } else {
        const status = data.building ? "BUILDING" : (data.result ?? "UNKNOWN");
        const duration = data.building
          ? formatDuration(Date.now() - data.timestamp) + " (so far)"
          : formatDuration(data.duration);
        console.log(`Build:    ${data.displayName ?? `#${data.number}`}`);
        console.log(`Status:   ${status}`);
        console.log(`Duration: ${duration}`);
        if (data.description) console.log(`Desc:     ${data.description}`);
        console.log(`URL:      ${data.url}`);
      }
      if (!data.building && data.result !== "SUCCESS") {
        process.exitCode = 1;
      }
    });

  program
    .command("set-build-description")
    .description("Set the description of a build")
    .argument("<name>", "job name")
    .argument("<build>", "build number")
    .argument("<description>", "description text")
    .action(async (name: string, build: string, description: string) => {
      const ctx = await resolveContext(program);
      await ctx.client.postForm(
        `${ctx.client.jobUrl(name)}/${build}/submitDescription`,
        { description },
      );
      console.log(`Build #${build} description updated.`);
    });

  program
    .command("set-build-display-name")
    .description("Set the display name of a build")
    .argument("<name>", "job name")
    .argument("<build>", "build number")
    .argument("<displayName>", "display name")
    .action(async (name: string, build: string, displayName: string) => {
      const ctx = await resolveContext(program);
      await ctx.client.postForm(
        `${ctx.client.jobUrl(name)}/${build}/configSubmit`,
        { displayName },
      );
      console.log(`Build #${build} display name updated.`);
    });

  program
    .command("delete-builds")
    .description("Delete builds")
    .argument("<name>", "job name")
    .argument("<builds>", "build numbers (comma-separated)")
    .action(async (name: string, builds: string) => {
      const ctx = await resolveContext(program);
      const jobUrl = ctx.client.jobUrl(name);
      const numbers = builds.split(",").map((s) => s.trim());
      await Promise.all(numbers.map((num) => ctx.client.post(`${jobUrl}/${num}/doDelete`)));
      console.log(`Deleted build(s): ${numbers.join(", ")}`);
    });

  program
    .command("list-changes")
    .description("List changes in a build")
    .argument("<name>", "job name")
    .argument("[build]", "build number (default: lastBuild)", "lastBuild")
    .action(async (name: string, build: string) => {
      const ctx = await resolveContext(program);
      const data = await ctx.client.getJson<{
        changeSet: {
          items: { msg: string; author: { fullName: string }; date?: string }[];
        };
      }>(
        `${ctx.client.jobUrl(name)}/${build}/api/json?tree=changeSet[items[msg,author[fullName],date]]`,
      );
      const items = data.changeSet?.items ?? [];
      if (ctx.json) {
        printOutput(items, true);
      } else if (items.length === 0) {
        console.log("No changes.");
      } else {
        for (const item of items) {
          console.log(`${item.author.fullName}: ${item.msg}`);
        }
      }
    });
}

interface BuildInfo {
  number: number;
  result: string | null;
  building: boolean;
  duration: number;
  estimatedDuration: number;
  timestamp: number;
  displayName: string | null;
  description: string | null;
  url: string;
}

export function formatDuration(ms: number): string {
  if (ms < 1000) return `${ms}ms`;
  const seconds = Math.floor(ms / 1000) % 60;
  const minutes = Math.floor(ms / 60000) % 60;
  const hours = Math.floor(ms / 3600000);
  if (hours > 0) return `${hours}h ${minutes}m ${seconds}s`;
  if (minutes > 0) return `${minutes}m ${seconds}s`;
  return `${seconds}s`;
}

File: cmd/builds.go
package cmd

import (
	"fmt"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/nwp/jenkins-cli/internal/console"
	"github.com/nwp/jenkins-cli/internal/output"
)

func newBuildsCmd() *cobra.Command {
	builds := &cobra.Command{
		Use:   "builds",
		Short: "Build management commands",
	}
	builds.AddCommand(
		newConsoleCmd(),
		newGetBuildCmd(),
		newWaitBuildCmd(),
		newListBuildsCmd(),
		newTailCmd(),
		newSetBuildDescriptionCmd(),
		newSetBuildDisplayNameCmd(),
		newDeleteBuildsCmd(),
		newListChangesCmd(),
	)
	return builds
}

func buildRef(args []string, idx int) string {
	if idx < len(args) && args[idx] != "" {
		return args[idx]
	}
	return "lastBuild"
}

func newConsoleCmd() *cobra.Command {
	var follow bool
	c := &cobra.Command{
		Use:   "console <name> [build]",
		Short: "Get console output for a build",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			jobURL := cl.JobURL(args[0])
			ref := buildRef(args, 1)
			if follow {
				return console.Follow(cl, jobURL, ref, os.Stdout)
			}
			text, err := console.GetText(cl, jobURL, ref)
			if err != nil {
				return err
			}
			fmt.Print(text)
			return nil
		},
	}
	c.Flags().BoolVar(&follow, "follow", false, "stream console output in real time")
	return c
}

func newGetBuildCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "get-build <name> [build]",
		Short: "Get build details",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			var b struct {
				Number      int    `json:"number"`
				Result      string `json:"result"`
				Building    bool   `json:"building"`
				Duration    int64  `json:"duration"`
				Timestamp   int64  `json:"timestamp"`
				DisplayName string `json:"displayName"`
			}
			path := fmt.Sprintf("%s/%s/api/json", cl.JobURL(args[0]), buildRef(args, 1))
			if err := cl.GetJSON(path, &b); err != nil {
				return err
			}
			if flagJSON {
				return output.Print(b, true)
			}
			status := b.Result
			if b.Building {
				status = "BUILDING"
			}
			ts := time.Unix(b.Timestamp/1000, 0).Local().Format(time.RFC3339)
			fmt.Printf("#%d  %s  %s  %s\n", b.Number, b.DisplayName, status, ts)
			fmt.Printf("Duration: %s\n", formatDuration(b.Duration))
			return nil
		},
	}
}

func newWaitBuildCmd() *cobra.Command {
	var timeoutSec int
	c := &cobra.Command{
		Use:   "wait-build <name> [build]",
		Short: "Wait for a build to complete",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			return waitForBuild(cl, cl.JobURL(args[0]), buildRef(args, 1), timeoutSec)
		},
	}
	c.Flags().IntVar(&timeoutSec, "timeout", 0, "timeout in seconds (0 = no timeout)")
	return c
}

func newListBuildsCmd() *cobra.Command {
	var limit int
	c := &cobra.Command{
		Use:   "list-builds <name>",
		Short: "List recent builds for a job",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			var result struct {
				Builds []struct {
					Number      int    `json:"number"`
					Result      string `json:"result"`
					Building    bool   `json:"building"`
					Duration    int64  `json:"duration"`
					Timestamp   int64  `json:"timestamp"`
					DisplayName string `json:"displayName"`
				} `json:"builds"`
			}
			tree := fmt.Sprintf("builds[number,result,building,duration,timestamp,displayName]{0,%d}", limit)
			path := cl.JobURL(args[0]) + "/api/json?tree=" + url.QueryEscape(tree)
			if err := cl.GetJSON(path, &result); err != nil {
				return err
			}
			if flagJSON {
				return output.Print(result.Builds, true)
			}
			for _, b := range result.Builds {
				status := b.Result
				if b.Building {
					status = "BUILDING"
				}
				ts := time.Unix(b.Timestamp/1000, 0).Local().Format("2006-01-02 15:04:05")
				fmt.Printf("#%d  %-10s  %s  %s\n", b.Number, status, formatDuration(b.Duration), ts)
			}
			return nil
		},
	}
	c.Flags().IntVar(&limit, "limit", 10, "maximum number of builds to show")
	return c
}

func newTailCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "tail <name> [build]",
		Short: "Stream new console output only (from current end position)",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			return console.Tail(cl, cl.JobURL(args[0]), buildRef(args, 1), os.Stdout)
		},
	}
}

func newSetBuildDescriptionCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "set-build-description <name> <build> <description>",
		Short: "Set the description of a build",
		Args:  cobra.ExactArgs(3),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			path := fmt.Sprintf("%s/%s/submitDescription", cl.JobURL(args[0]), args[1])
			resp, err := cl.PostForm(path, url.Values{"description": {args[2]}})
			if err != nil {
				return err
			}
			drainClose(resp)
			fmt.Println("Description updated.")
			return nil
		},
	}
}

func newSetBuildDisplayNameCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "set-build-display-name <name> <build> <displayName>",
		Short: "Set the display name of a build",
		Args:  cobra.ExactArgs(3),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			path := fmt.Sprintf("%s/%s/configSubmit", cl.JobURL(args[0]), args[1])
			resp, err := cl.PostForm(path, url.Values{"displayName": {args[2]}})
			if err != nil {
				return err
			}
			drainClose(resp)
			fmt.Println("Display name updated.")
			return nil
		},
	}
}

func newDeleteBuildsCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete-builds <name> <builds>",
		Short: "Delete builds (comma-separated build numbers)",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			jobURL := cl.JobURL(args[0])
			for _, num := range strings.Split(args[1], ",") {
				num = strings.TrimSpace(num)
				path := fmt.Sprintf("%s/%s/doDelete", jobURL, num)
				resp, err := cl.PostEmpty(path)
				if err != nil {
					return fmt.Errorf("delete build %s: %w", num, err)
				}
				drainClose(resp)
				fmt.Printf("Build #%s deleted.\n", num)
			}
			return nil
		},
	}
}

func newListChangesCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "list-changes <name> [build]",
		Short: "List changes/commits in a build",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			cl, err := resolveClient()
			if err != nil {
				return err
			}
			var result struct {
				ChangeSets []struct {
					Items []struct {
						CommitID string `json:"commitId"`
						Msg      string `json:"msg"`
						Author   struct {
							FullName string `json:"fullName"`
						} `json:"author"`
					} `json:"items"`
				} `json:"changeSets"`
			}
			path := fmt.Sprintf("%s/%s/api/json?tree=changeSets[items[commitId,msg,author[fullName]]]",
				cl.JobURL(args[0]), buildRef(args, 1))
			if err := cl.GetJSON(path, &result); err != nil {
				return err
			}
			if flagJSON {
				return output.Print(result.ChangeSets, true)
			}
			for _, cs := range result.ChangeSets {
				for _, item := range cs.Items {
					id := item.CommitID
					if len(id) > 8 {
						id = id[:8]
					}
					fmt.Printf("%s  %s  %s\n", id, item.Author.FullName, item.Msg)
				}
			}
			return nil
		},
	}
}

// formatDuration formats a Jenkins build duration (milliseconds) as a human-readable string.
func formatDuration(ms int64) string {
	if ms < 1000 {
		return fmt.Sprintf("%dms", ms)
	}
	secs := ms / 1000
	if secs < 60 {
		return fmt.Sprintf("%ds", secs)
	}
	mins := secs / 60
	secs = secs % 60
	if mins < 60 {
		return fmt.Sprintf("%dm %ds", mins, secs)
	}
	hours := mins / 60
	mins = mins % 60
	return fmt.Sprintf("%dh %dm %ds", hours, mins, secs)
}
